src/lld/moviebooking/class_diagram.puml

```diff
@@ -1,3 +1,323 @@
+        + DBMovieCityWriteDao()
+        + create(movie: Movie): Movie
+        + update(id: long, movie: Movie): Movie
+        + delete(id: long): boolean
+    }
+    class DBTheatreReadDao<Theatre> implements ReadDao{
+        - connection: Connection
+        + DBTheatreReadDao()
+        + findAll(): List<Theatre>
+        + findById(id: long): List<Theatre>
+    }
+    class DBTheatreWriteDao<Theatre> implements WriteDao{
+        - connection: Connection
+        + DBTheatreWriteDao()
+        + create(movie: Theatre): Theatre
+        + update(id: long, movie: Theatre): Theatre
+        + delete(id: long): boolean
+    }
+    class DBTheatreGroupReadDao<TheatreGroup> implements ReadDao{
+        - connection: Connection
+        + DBTheatreGroupReadDao()
+        + findAll(): List<TheatreGroup>
+        + findById(id: long): TheatreGroup
+    }
+    class DBTheatreGroupWriteDao<TheatreGroup> implements WriteDao{
+        - connection: Connection
+        + DBTheatreGroupWriteDao()
+        + create(theatreGroup: TheatreGroup): TheatreGroup
+        + update(id: long, theatreGroup: TheatreGroup): TheatreGroup
+        + delete(id: long): boolean
+    }
+    class DBScreenReadDao<Screen> implements ReadDao{
+        - connection: Connection
+        + DBScreenReadDao()
+        + findAll(): List<Screen>
+        + findById(id: long): Screen
+    }
+    class DBScreenWriteDao<Screen> implements ReadDao{
+        - connection: Connection
+        + DBScreenWriteDao()
+        + create(screen: Screen): Screen
+        + update(id: long, screen: Screen): Screen
+        + delete(id: long): boolean
+    }
+    class DBSeatReadDao<Seat> implements ReadDao{
+        - connection: Connection
+        + DBSeatReadDao()
+        + findById(id: long): Seat
+        + findAll(): List<Seat>
+    }
+    class DBSeatWriteDao<Seat> implements WriteDao{
+        - connection: Connection
+        + DBSeatWriteDao()
+        + create(seat: Seat): Seat
+        + update(id: long, seat: Seat): Seat
+        + delete(id: long): boolean
+    }
+    class DBShowReadDao<Show> implements ReadDao{
+        - connection: Connection
+        + DBShowReadDao()
+        + findAll(): List<Show>
+        + findById(id: long): Show
+    }
+    class DBShowWriteDao<Show> implements WriteDao{
+        - connection: Connection
+        + DBShowWriteDao()
+        + create(show: Show): Seat
+        + update(id: long, show: Show): Show
+        + delete(id: long): boolean
+    }
+}
+
+package com.movie_booking.dao.mapping {
+    interface MovieCityReadDao{
+        + findMoviesByCity(cityId: long): List<Movie>
+        + findCitiesByMovie(movieId: long): List<City>
+    }
+    interface MovieCityWriteDao{
+        + addMoviesToCity(movieId: long, cityId: long)
+    }
+    'FR1: Get all movies by city query uses these DAO
+    class DBMovieCityReadDao implements MovieCityReadDao{
+        - connection: Connection
+        + DBMovieCityReadDao(connection)
```

Comment:
include main.dao which are to be used here? for example: DBMovieCityReadDao might require movieReadDao, cityDao

Reply:
I have few queries on this? let me find any recommendation are there for mutiple different type interface extension to one common interface?
My ReadDAO is generic interface with one type bound, let say DBMovieReadDao<Movie> implements generic ReadDao<T> but here the situation is DBMovieCityReadDAO require ReadDao<Movie> and ReadDao<City> combined interface should be implemented

Reply:
`public interface ListPrinter<T> {
    void printList(List<T> list);
}

public interface MultiTypeListPrinter<T, U> extends ListPrinter<T>, ListPrinter<U> {
    // empty interface
}

public class StringIntegerListPrinter implements MultiTypeListPrinter<String, Integer> {
    public void printList(List<String> list) {
        for (String item : list) {
            System.out.println("Printing String: " + item);
        }
    }

    public void printList(List<Integer> list) {
        for (Integer item : list) {
            System.out.println("Printing Integer: " + item);
        }
    }
}`